### PR TITLE
feat: upgrade argo-workflows to v3.1.0. Part of #5718

### LINKED ISSUE #5718: [feature] default to the emissary executor
#### Update - 2021.9.13
edited by @Bobgy 

KFP 1.7.0 is released.
documentation is live: https://www.kubeflow.org/docs/components/pipelines/installation/choose-executor/
choosing an executor will be an option starting from KFP 1.7.

#### Update - 2021.8.6
edited by @Bobgy 

Discussed with @kramachandran and decided to delay the default change for at least one KFP minor version, because argo emissary executor is still in Alpha state. There can be other corner cases that we haven't found, so we should give people some time to test it out.

Therefore, I'd like to propose the following:
* still default to docker executor in 1.7.0 release
* provide emissary executor as an option and document how to migrate
* collect feedback from people who starts to try emissary executor

TODOs:
* [x] https://github.com/kubeflow/website/pull/2857
* [x] Make emissary available in installations
  * [x] #6294
  * [x] Make env/platform-agnostic-emissary
* [x] #6252
* [x] Upgrade kfp-ci test infra to argo v3.1.1 https://github.com/kubeflow/testing/commit/03c625828ac9cc13283bd226371280b0f5704c09

Issues discovered when testing argo v3.1.0 with KFP:
* [x] https://github.com/kubeflow/pipelines/issues/5930
* [x] https://github.com/kubeflow/pipelines/issues/5937 (fixed in v3.1.1)
* [x] https://github.com/kubeflow/pipelines/issues/5944 (fixed in v3.1.1)
* [x] https://github.com/argoproj/argo-workflows/issues/6238

=== the following are the original proposal ===

### What feature would you like to see?

Update argo to 3.11 such that we can use the emissary executor as decided by @bobgy in https://github.com/kubeflow/pipelines/pull/4645#issuecomment-845691359

### What is the use case or pain point?

The Docker, Kubelet, PNS and K8sapi executors have severe limitations that will be solved by Argo 3.1 and the emissary executor.

Here is an overview https://argoproj.github.io/argo-workflows/workflow-executors/

In a nutshell docker breaks security completely and is incompatible with kubernetes 1.19
https://github.com/kubeflow/pipelines/issues/5714 Kubernetes is also moving to containerd anyway and a lot of users want proper non-docker support for a long time https://github.com/kubeflow/pipelines/issues/1654

Kubelet needs some configuration and has the same limitation as k8sapi and rootless PNS: "Output artifacts must be saved on volumes (e.g. emptyDir) and not the base image layer (e.g. /tmp)" so it breaks the leightweight python components if there is no volume mounted or the output directory is inside the base layer https://github.com/kubeflow/pipelines/pull/4645

One can use a PNS with root rights (PTRACE and CHROOT) to circumvent this, but this obviously breaks the security again.

 So only the emissary executor https://argoproj.github.io/argo-workflows/workflow-executors/#emissary-emissary is feasible which does more ore less what i proposed in https://github.com/kubeflow/pipelines/pull/4645 and what was started by Argos @alexec in https://github.com/argoproj/argo-workflows/pull/4766

The Argo update has been done before in https://github.com/kubeflow/pipelines/pull/5266 and https://github.com/kubeflow/pipelines/pull/4693 by @xinbinhuang @Bobgy 

### Is there a workaround currently?

```
# Emmissary executor, needs argo 3.1+
# Sadly visualizations are broken(26.04.2021) for k8sapi and emissary

kubectl -n kubeflow patch configmap/workflow-controller-configmap --type='json' -p='[{"op": "replace", "path": "/data/containerRuntimeExecutor", "value": "emissary"}]'

kubectl edit deployment/workflow-controller -n kubeflow
...
    spec:
      containers:
      - env:
        - name: "LEADER_ELECTION_DISABLE"
       	  value: true
      - args:
	- --configmap
        - workflow-controller-configmap
        - --executor-image
        - docker.io/argoproj/argoexec:latest
        command:
        - workflow-controller
        image: docker.io/argoproj/workflow-controller:latest    
        env:
	- name: "LEADER_ELECTION_DISABLE"
          value: "true"
...
kubectl -n kubeflow patch configmap/workflow-controller-configmap --type='json' -p='[{"op": "replace", "path": "/data/containerRuntimeExecutor", "value": "emissary"}]'
kubectl rollout restart deployment workflow-controller -n kubeflow

or manually patching the sdk and using k8sapi as done in https://github.com/kubeflow/pipelines/pull/4645

Or use kfp-tekton
```

---

Love this idea? Give it a 👍. We prioritize fulfilling features with the most 👍.

### PR DESCRIPTION
**Pull request by Bobgy**
**Description of your changes:**
Part of #5718 

### Changes
* upgrade argo go module
* upgrade argo manifests
* update licenses for go binaries
* update documentation for how to upgrade argo
* fix everything after the upgrade

**Checklist:**
- [x] The title for your pull request (PR) should follow our title convention. [Learn more about the pull request title convention used in this repository](https://github.com/kubeflow/pipelines/blob/master/CONTRIBUTING.md#pull-request-title-convention).

### CODE REVIEW

File: manifests/kustomize/third-party/argo/installs/namespace/cluster-scoped/kustomization.yaml [hunk 1]
```diff
@@ -1,4 +1,5 @@
 apiVersion: kustomize.config.k8s.io/v1beta1
 kind: Kustomization
 bases:
-- ../../../upstream/manifests/base/crds
+# Minimal CRDs omit schema validation, recommended for production cluster.
```
Comment (NikeNano) [inline manifests/kustomize/third-party/argo/installs/namespace/cluster-scoped/kustomization.yaml:4]: to learn: what schema validation is referred to here?

File: manifests/kustomize/third-party/argo/upstream/manifests/quick-start/base/artifact-repositories-configmap.yaml [hunk 2]
```diff
@@ -11,6 +19,16 @@ data:
...
+      bucket: my-bucket
+      endpoint: minio:9000
+      insecure: true
+      accessKeySecret:
+        name: my-minio-cred
+        key: accesskey
+      secretKeySecret:
+        name: my-minio-cred
+        key: secretkey
+  empty: ""
```
Comment (NikeNano) [inline manifests/kustomize/third-party/argo/upstream/manifests/quick-start/base/artifact-repositories-configmap.yaml:34]: new line?

File: manifests/kustomize/third-party/argo/upstream/manifests/quick-start/base/kubelet-executor-clusterrole.yaml [hunk 3]
```diff
@@ -0,0 +1,12 @@
...
+metadata:
+  name: kubelet-executor
+rules:
+  # This allows the kubelet executor.
+  - apiGroups:
+      - ""
+    resources:
+      - nodes/proxy
+    verbs:
+      - get
```
Comment (NikeNano) [inline manifests/kustomize/third-party/argo/upstream/manifests/quick-start/base/kubelet-executor-clusterrole.yaml:12]: newline?

File: manifests/kustomize/third-party/argo/upstream/manifests/quick-start/base/minio/minio-deploy.yaml [hunk 4]
```diff
@@ -0,0 +1,42 @@
...
+              path: /minio/health/ready
+              port: 9000
+            initialDelaySeconds: 5
+            periodSeconds: 10
+          livenessProbe:
+            httpGet:
+              path: /minio/health/live
+              port: 9000
+            initialDelaySeconds: 5
+            periodSeconds: 10
```
Comment (NikeNano) [inline manifests/kustomize/third-party/argo/upstream/manifests/quick-start/base/minio/minio-deploy.yaml:42]: nit: newline?

File: manifests/kustomize/third-party/argo/upstream/manifests/quick-start/base/prometheus/prometheus-deployment.yaml [hunk 5]
```diff
@@ -0,0 +1,37 @@
...
+          args:
+            - --config.file=/config/prometheus.yaml
+          volumeMounts:
+          - name: config
+            mountPath: /config
+      volumes:
+        - name: config
+          configMap:
+            name: prometheus-config
+
```
Comment (NikeNano) [inline manifests/kustomize/third-party/argo/upstream/manifests/quick-start/base/prometheus/prometheus-deployment.yaml:36]: nit: remove one line?

File: manifests/kustomize/third-party/argo/installs/namespace/cluster-scoped/kustomization.yaml (hunk 1 above)
Comment (Bobgy) [reply manifests/kustomize/third-party/argo/installs/namespace/cluster-scoped/kustomization.yaml:4]: See https://github.com/argoproj/argo-workflows/tree/master/manifests/base/crds/full

> These CRDs have full schema validation. As a result, they are large and probably not suitable to be used in your cluster.

File: manifests/kustomize/third-party/argo/upstream/manifests/quick-start/base/artifact-repositories-configmap.yaml (hunk 2 above)
Comment (Bobgy) [reply manifests/kustomize/third-party/argo/upstream/manifests/quick-start/base/artifact-repositories-configmap.yaml:34]: Note these files are pulled from argo without touch, so I'd recommend avoid making any changes.

File: manifests/kustomize/third-party/argo/installs/namespace/cluster-scoped/kustomization.yaml (hunk 1 above)
Comment (NikeNano) [reply manifests/kustomize/third-party/argo/installs/namespace/cluster-scoped/kustomization.yaml:4]: Thanks!

File: manifests/kustomize/third-party/argo/upstream/manifests/quick-start/base/artifact-repositories-configmap.yaml (hunk 2 above)
Comment (NikeNano) [reply manifests/kustomize/third-party/argo/upstream/manifests/quick-start/base/artifact-repositories-configmap.yaml:34]: Yes that makes sense.

**Comment by Bobgy** [comment]
I got an error `standard_init_linux.go:211: exec user process caused "permission denied"` from the frontend integration test pod in kubeflow-pipeline-e2e-test.

**Comment by Bobgy** [comment]
/hold
Filed upstream issue for using argo emissary executor with rootless container: https://github.com/argoproj/argo-workflows/issues/6238.

The test failure is at selenium image: https://github.com/kubeflow/pipelines/blob/76f8b3e9acf09bdc5a7e7729b0dfa64dd51b922a/test/frontend-integration-test/Dockerfile#L4. It runs as non-root.

**Comment by Bobgy** [comment]
Potential solutions
* maybe we can upgrade selenium image to latest, so that it configures user as a numeric value: https://github.com/SeleniumHQ/docker-selenium/blob/selenium-3/StandaloneChrome/Dockerfile this is required to configure rootless pods in Kubernetes security context https://kubernetes.io/docs/tasks/configure-pod-container/security-context/#set-the-security-context-for-a-pod

**Comment by Bobgy** [comment]
Because the remaining problems are related to switching to emissary executor. I'd like to separate the PR into two:
1. Upgrade argo to v3.1.0
2. Switch to emissary executor

**Comment by Bobgy** [comment]
/unhold
this PR only upgrades argo to v3.1.0

**Comment by Bobgy** [comment]
/approve

self approving the low risk upgrade PR, will request further review for switching to emissary executor
